### DISCUSSION TITLE
Temporary file .db-wal

### DISCUSSION THREAD

**Discussion by BPilot253** [pocketbase/pocketbase#1265] [Q&A] [answered]
I noticed a .db-wal file is created alongside the .db file. I found out that it's a temporary file created by SQLite for rollback if something fails. I use Storj for S3 storage since they give you 150GB storage for free and I have ~40GB free storage on my VPS. My question is: what will happen when I have over 40GB of data? Will it only back up the most recent files or all? Also, is that same .db-wal file used when S3 connection fails?

_(п.п. радвам се, че този проект е направен от друг българин, наистина е много полезен и страхотно направен, поздравления 🙂)_

**Comment by ganigeorgiev** [comment, marked as answer]
The `-wal` file is not for "rollback". It contains executed writes/transactions that haven't been applied/transferred to the main db file yet (which happens during [checkpointing](https://www.sqlite.org/wal.html#ckpt) and usually is executed automatically after some threshold or on sqlite close_).

Those 2 temp files (if they exist) needs to be included in the backup. Technically the `-shm` index is not required since it is used mainly for coordinating access between concurrent clients, but the `-wal` file is critical (if exist) since, depending at what stage the process was abruptly terminated, without it you may end up with missing writes or even corrupted main db file.

> what will happen when I have over 40GB of data?

You'll get a "disk is full" or other similar error when trying to insert/update something in the database. Note that removing entries from the database doesn't shrink the db file and you'll need to call `VACUUM` in order to reclaim the freed disk space.

> Will it only back up the most recent files or all?
> Also, is that same .db-wal file used when S3 connection fails?

I'm not sure that I understand the questions.

Only the uploaded user files can be stored on S3. The database needs to be mounted on the server where the PocketBase executable is running, aka. on the local filesytem (there are options for various vfs but for now PocketBase doesn't make use of them). If you want to backup (and eventually restore) your database to/from a S3 storage you can use [Litestream](https://litestream.io/) or something more generic like [rclone](https://github.com/rclone/rclone).

**Reply by BPilot253** [reply]
Oh, I must've misunderstood then. So, the -wal file only stores the most recent writes/transactions that haven't yet passed the "checkpoint" aka. that haven't yet been uploaded to the connected S3 storage? That's where I was confused and the the question you didn't understand is now irrelevant. Thanks a lot for the detailed explanation.

**Reply by ganigeorgiev** [reply]
> that haven't yet been uploaded to the connected S3 storage

The main database file is always stored on the disk.

The "_Use S3 storage_" option in the Admin UI is only for the user uploaded files, eg. images, documents, or anything else that was attached to a collection `file` field. In the database we store only the name of the uploaded file, but the file content itself can be stored either locally (`pb_data/storage`) or remotely (in a S3 bucket).

**Reply by BPilot253** [reply]
Oh. Okay, in that case I completely misunderstood everything.
